### PR TITLE
[FinalizePublicClassConstantRector] Ignore final classes

### LINKED ISSUE #6964: [FinalizePublicClassConstantRector] Ignore final classes
# Feature Request

| Subject        | Details              |
| :------------- | :--------------------|
| Rector version | last dev-master      |
| Installed as   | composer dependency  |

## Minimal PHP Code Causing Issue

See https://getrector.org/demo/1ec7d0d7-7bb2-68cc-b5e0-11dedf8f8668

```php
<?php

final class DemoFile
{
    public const TEST_CONSTANT = 'test';
}
```

### Responsible rules

* `FinalizePublicClassConstantRector`

## Expected Behavior

The rector should ignore classes that are marked as final as the class cannot be extended anyway, thus making the constants final is redundant. 

I expect the rector to not change the file at all if it is final.

### PR DESCRIPTION
**Pull request by reypm**
Fixes issue with Rector changing `const` definition to final when the class is already defined as `final`
Fixes https://github.com/rectorphp/rector/issues/6964 
Closes https://github.com/rectorphp/rector-src/pull/1720

### CODE REVIEW

File: rules/Php81/Rector/ClassConst/FinalizePublicClassConstantRector.php [hunk 1]
```diff
@@ -60,6 +60,13 @@ public function getNodeTypes(): array
      */
     public function refactor(Node $node): ?Node
     {
+        /** @var Node\Stmt\Class_ $parentClass */
+        $parentClass = $this->betterNodeFinder->findParentByTypes($node, [Node\Stmt\Class_::class]);
+
+        if ($parentClass->isFinal()) {
```
Comment (samsonasik) [inline rules/Php81/Rector/ClassConst/FinalizePublicClassConstantRector.php]: when parentClass is not a Class_, it need to be skipped early:

```suggestion
        $parentClass = $this->betterNodeFinder->findParentType($node, Node\Stmt\Class_::class);

        if (! $parentClass instanceof Node\Stmt\Class_) {
            return null;
        }

        if ($parentClass->isFinal()) {
```

Comment (reypm) [reply rules/Php81/Rector/ClassConst/FinalizePublicClassConstantRector.php]: Maybe I should check on `null` instead? 🤔

File: rules/Php81/Rector/ClassConst/FinalizePublicClassConstantRector.php [hunk 2]
```diff
@@ -60,6 +60,16 @@ public function getNodeTypes(): array
      */
     public function refactor(Node $node): ?Node
     {
+        $parentClass = $this->betterNodeFinder->findParentByTypes($node, [Node\Stmt\Class_::class]);
```
Comment (samsonasik) [inline rules/Php81/Rector/ClassConst/FinalizePublicClassConstantRector.php]: findParentType can be used:

```suggestion
        $parentClass = $this->betterNodeFinder->findParentType($node, Node\Stmt\Class_::class);
```

**Comment by samsonasik** [comment]
Thank you @reypm

**Comment by TomasVotruba** [comment]
Thanks :+1:

**Pull request merged**